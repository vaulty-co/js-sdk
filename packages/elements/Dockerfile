FROM node:current-alpine3.11

WORKDIR /app

COPY packages ./packages

COPY package.json yarn.lock ./

# Install production dependecies

RUN yarn install --frozen-lockfile --production

# Remove unused packages

RUN cd packages && ls -1 | grep -v "elements" | xargs rm -rf

# Start @vaulty/elements server

CMD cd packages/elements && yarn start

EXPOSE 3001
