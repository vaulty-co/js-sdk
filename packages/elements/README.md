# JS SDK elements (fields and its controllers)

It creates content inside of iframe by get params (for example, an input)

## Technical requirements

 - Node v10.13+
 - Yarn 1.19+

## Initialize repository

```sh
yarn install
```

## Start Dev version of elements controller

```sh
yarn start
```

## Run linting

```sh
yarn lint
```

## Run tests

```sh
yarn test
```

## Build library

```sh
yarn build
```
