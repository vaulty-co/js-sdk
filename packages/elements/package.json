{
  "name": "@vaulty/elements",
  "version": "0.1.0",
  "description": "Vaulty Elements (fields and its controllers)",
  "scripts": {
    "dev": "rimraf ./devTmp && NODE_ENV=development node ./server",
    "prod": "yarn build && NODE_ENV=production node ./server",
    "start": "NODE_ENV=production node ./server",
    "lint": "eslint --cache ./src ./internals ./server",
    "lint:ci": "eslint ./src ./internals ./server",
    "test": "jest --config=./internals/tests/jest.config.js --passWithNoTests",
    "build": "rimraf ./build && NODE_ENV=production rollup -c ./internals/rollup/rollup.config.js"
  },
  "author": "Aliaksei Simanchyk <alex@vaulty.co>",
  "license": "ISC",
  "devDependencies": {
    "@babel/core": "7.10.2",
    "@babel/plugin-proposal-nullish-coalescing-operator": "^7.10.1",
    "@babel/plugin-proposal-optional-chaining": "^7.10.1",
    "@babel/preset-env": "^7.10.2",
    "@rollup/plugin-alias": "^3.1.1",
    "@rollup/plugin-commonjs": "12.0.0",
    "@rollup/plugin-image": "^2.0.5",
    "@rollup/plugin-node-resolve": "^8.0.1",
    "babel-jest": "^26.0.1",
    "babel-plugin-transform-define": "^2.0.0",
    "jest": "^26.0.1",
    "morgan": "^1.10.0",
    "node-sass": "^4.14.1",
    "reload": "^3.0.5",
    "rimraf": "^3.0.2",
    "rollup": "^2.14.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-gzip": "^2.5.0",
    "rollup-plugin-postcss": "^3.1.2",
    "rollup-plugin-terser": "^6.1.0",
    "rollup-plugin-visualizer": "^4.0.4",
    "zlib": "^1.0.5"
  },
  "dependencies": {
    "@vaulty/common": "0.1.0",
    "card-validator": "^6.2.0",
    "express": "^4.17.1",
    "express-static-gzip": "^2.0.6",
    "history": "^4.10.1",
    "imask": "^6.0.5",
    "object-path": "^0.11.4",
    "pug": "^2.0.4",
    "url-search-params-polyfill": "^8.1.0"
  }
}
