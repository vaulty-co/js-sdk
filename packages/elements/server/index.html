<!doctype html>
<html>
    <head>
        <title>Reload Express Sample App</title>
    </head>
    <body>
        <div id="app"></div>
        <script src="js-sdk-elements.min.js"></script>
        <!-- All you have to do is include the reload script and have it be on every page of your project -->
        <!-- You do not create this route, reload creates it for you automatically -->
        <script src="/reload/reload.js"></script>
    </body>
</html>
