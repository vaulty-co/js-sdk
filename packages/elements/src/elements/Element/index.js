import invariant from 'invariant';

import { NODE_TYPES } from '../constants/nodeTypes';

const STATUSES = {
  INIT: 'init',
  DESTROYED: 'destroyed',
};

/**
 * Base class for Elements
 * @class
 * @extends {VaultyElementsInstance}
 */
class Element {
  /**
   * @param {boolean} condition
   * @param {string} message
   */
  static invariant(condition, message) {
    invariant(condition, `[${this.name}] ${message}`);
  }

  static get STATUSES() {
    return STATUSES;
  }

  /**
   * @param {HTMLElement} node
   */
  constructor(node) {
    this.constructor.invariant(
      typeof node === 'object' && node.nodeType === NODE_TYPES.ELEMENT_NODE,
      'Node for element should be a HTMLElement',
    );

    this.status = STATUSES.INIT;
    this.node = node;
  }

  /**
   * Mount element
   * @param {HTMLElement} node
   */
  mount(node) {
    this.appendTo(node);
  }

  /**
   * Unmount element
   */
  unmount() {
    this.destroy();
  }

  /**
   * Append node to parent element
   * @param {HTMLElement} parent - parent node
   */
  appendTo(parent) {
    if (this.status === STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Element is destroyed and can not be append to a parent node.',
      );
      return;
    }

    this.constructor.invariant(
      typeof parent === 'object' && parent.nodeType === NODE_TYPES.ELEMENT_NODE,
      'Parent node should be a HTMLElement node',
    );

    this.parent = parent;
    parent.appendChild(this.node);
  }

  /**
   * Destroy element and remove its from parent, if it is specified
   */
  destroy() {
    if (this.status === STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Element is destroyed and can not be destroyed again.',
      );
      return;
    }

    if (this.parent) {
      this.parent.removeChild(this.node);
    }
    this.status = STATUSES.DESTROYED;
    this.node = null;
    this.parent = null;
  }
}

export default Element;
export {
  Element,
};
