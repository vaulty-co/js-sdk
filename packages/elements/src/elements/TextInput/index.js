import styles from './styles.scss';

/**
 * It renders simple text input node
 * @class
 */
class TextInput {
  constructor() {
    const node = document.createElement('input');
    node.setAttribute('type', 'text');
    node.className = styles.input;

    this.node = node;
  }

  /**
   * Append node to parent element
   * @param {HTMLElement} parent - parent node
   */
  appendTo(parent) {
    this.parent = parent;
    parent.appendChild(this.node);
  }

  /**
   * Destroy input element and remove its from parent, if it is specified
   */
  destroy() {
    if (this.parent) {
      this.parent.removeChild(this.node);
    }
    this.node = null;
    this.parent = null;
  }
}

export default TextInput;
export {
  TextInput,
};
