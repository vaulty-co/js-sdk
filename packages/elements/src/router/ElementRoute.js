import { Route } from './Route';
import { Element } from '../elements/Element';

/**
 * Route for Element
 */
class ElementRoute extends Route {
  /**
   * Render route
   * @returns {Element}
   */
  render() {
    const element = this.renderFunction();
    this.constructor.invariant(
      (element instanceof Element),
      `Element route "${this.name}" does not return Element in render function.`,
    );
    return element;
  }
}

export default ElementRoute;
export {
  ElementRoute,
};
