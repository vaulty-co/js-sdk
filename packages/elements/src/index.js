import 'url-search-params-polyfill';

import './styles';
import { CONSTANTS } from './constants';
import { Router } from './router';
import { inputTextRoute } from './elements/TextInput/route';
import { formRoute } from './controllers/Form/route';

const appNode = document.getElementById(CONSTANTS.APP_ID);
const router = new Router(appNode);
router.register(inputTextRoute);
router.register(formRoute);
router.render();
