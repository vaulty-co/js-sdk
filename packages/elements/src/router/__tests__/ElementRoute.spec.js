import { Element } from 'vaulty-js-sdk-elements/src/elements/Element';
import { ElementRoute } from '../ElementRoute';

describe('ElementRoute', () => {
  let route;
  let renderFunction;

  beforeEach(() => {
    renderFunction = jest.fn(() => {
      const input = document.createElement('input');
      input.type = 'text';
      return new Element(input);
    });
    route = new ElementRoute(
      'inputText',
      {
        field: 'input',
        type: 'text',
      },
      renderFunction,
    );
  });

  describe('#render', () => {
    it('should run render function', () => {
      route.render();

      expect(renderFunction).toBeCalled();
    });

    it('should throw exception, when it does not return Element instance', () => {
      route = new ElementRoute(
        'inputText',
        {
          field: 'input',
          type: 'text',
        },
        () => null,
      );

      expect(() => {
        route.render();
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
