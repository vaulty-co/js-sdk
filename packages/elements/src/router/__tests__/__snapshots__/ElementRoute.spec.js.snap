// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementRoute #render should throw exception, when it does not return Element instance 1`] = `"[ElementRoute] Element route \\"inputText\\" does not return Element in render function."`;
